compiler/rustc_trait_selection/src/solve/infcx.rs

```diff
@@ -0,0 +1,435 @@
+use std::ops::Deref;
+
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir::def_id::{DefId, LocalDefId};
+use rustc_infer::infer::canonical::query_response::make_query_region_constraints;
+use rustc_infer::infer::canonical::{
+    Canonical, CanonicalExt as _, CanonicalVarInfo, CanonicalVarValues,
+};
+use rustc_infer::infer::{
+    BoundRegionConversionTime, InferCtxt, RegionVariableOrigin, SubregionOrigin, TyCtxtInferExt,
+};
+use rustc_infer::traits::solve::Goal;
+use rustc_infer::traits::util::supertraits;
+use rustc_infer::traits::{ObligationCause, Reveal};
+use rustc_middle::ty::fold::TypeFoldable;
+use rustc_middle::ty::{self, Ty, TyCtxt, TypeVisitableExt as _};
+use rustc_span::{ErrorGuaranteed, Span, DUMMY_SP};
+use rustc_type_ir::relate::Relate;
+use rustc_type_ir::solve::{Certainty, NoSolution, SolverMode};
+
+use crate::traits::coherence::trait_ref_is_knowable;
+use crate::traits::specialization_graph;
+
+#[repr(transparent)]
+pub struct SolverDelegate<'tcx>(InferCtxt<'tcx>);
```

Comment:
The reason I'm doing this is because I didn't want to something like:

```
struct SolverDelegate<'a, 'tcx>(&'a InferCtxt<'tcx>);
```

because then we'd be going through two layers of indirection.